File: JavaLibrary/src/main/java/com/flat/simplex/parser/logic/line/LineGroup.java
package com.flat.simplex.parser.logic.line;

import com.flat.simplex.lexer.Key;
import com.flat.simplex.lexer.Token;
import com.flat.simplex.parser.logic.Block;
import com.flat.simplex.parser.logic.Context;

import java.util.ArrayList;

public class LineGroup extends LineValue {

    public enum Type {
        Single, Prefix, Postfix, Middle, Ternary
    }

    private LineValue lineOne;
    private LineValue lineTwo;
    private LineValue lineTree;
    private LineOp lineOp;
    private LineOp lineElse;
    private Type type;
    private ArrayList<Line> lines = new ArrayList<>();

    public LineGroup(LineValue singleLine) {
        super(singleLine.getParent(), singleLine.getToken());
        this.lineOne = singleLine;
        this.type = Type.Single;
        lines.add(lineOne);
    }

    public LineGroup(LineOp leftOp, LineValue line) {
        super(leftOp.getParent(), leftOp.getToken());
        this.lineOp = leftOp;
        this.lineOne = line;
        this.type = Type.Prefix;
        lines.add(lineOp);
        lines.add(lineOne);
    }

    public LineGroup(LineValue line, LineOp rightOp) {
        super(line.getParent(), line.getToken());
        this.lineOp = rightOp;
        this.lineOne = line;
        this.type = Type.Postfix;
        lines.add(lineOne);
        lines.add(lineOp);
    }

    public LineGroup(LineValue leftLine, LineOp lineOp, LineValue rightLine) {
        super(leftLine.getParent(), leftLine.getToken());
        this.lineOne = leftLine;
        this.lineOp = lineOp;
        this.lineTwo = rightLine;
        this.type = Type.Middle;
        lines.add(lineOne);
        lines.add(lineOp);
        lines.add(lineTwo);
    }

    public LineGroup(LineValue leftLine, LineOp lineOp, LineValue rightLine, LineOp lineElse, LineValue lineTree) {
        super(leftLine.getParent(), leftLine.getToken());
        this.lineOne = leftLine;
        this.lineOp = lineOp;
        this.lineTwo = rightLine;
        this.lineElse = lineElse;
        this.lineTree = lineTree;
        this.type = Type.Ternary;
        lines.add(leftLine);
        lines.add(lineOp);
        lines.add(lineTwo);
        lines.add(lineElse);
        lines.add(lineTree);
    }

    public Context getContext() {
        return parent.getContext();
    }

    public Block getParent() {
        return parent;
    }

    public Type getType() {
        return type;
    }

    public LineValue getLineOne() {
        return lineOne;
    }

    public LineValue getLineTwo() {
        return lineTwo;
    }

    public LineValue getLineTree() {
        return lineTree;
    }

    public LineOp getLineOp() {
        return lineOp;
    }

    public LineOp getLineElse() {
        return lineElse;
    }

    public ArrayList<Line> getLines() {
        return lines;
    }

    @Override
    public ReturnType getReturnType() {
        if (type == Type.Middle && lineOp != null) {
            if (lineOp.getKey() == Key.And || lineOp.getKey() == Key.Or) {
                return ReturnType.Boolean;
            } else if (lineOp.getPrecedence() == Key.Set.op) {
                return ReturnType.Void;
            }
        }
        return super.getReturnType();
    }

    @Override
    public String toString() {
        if (type == Type.Single) return "(" + lineOne + ")";
        if (type == Type.Prefix) return  "(" + lineOp + " " + lineOne + ")";
        if (type == Type.Postfix) return  "(" + lineOne + " " + lineOp + ")";
        if (type == Type.Middle) return  "(" + lineOne + " " + lineOp + " " + lineTwo + ")";
        if (type == Type.Ternary) return  "(" + lineOne + " " + lineOp + " " + lineTwo + " " + lineElse + " " + lineTree + ")";
        return super.toString();
    }
}

File: SimplexApp/simplex/lang/parser/logic/line/LineGroup.js
var Key, LineOp, LineCall, Error;
const LineValue = require("simplex/lang/parser/logic/line/LineValue.js");

class LineGroup extends LineValue {

    constructor(...args$) {
        super(args$[0].getParent(), args$[0].getToken());
        this.lines = [];
        switch (args$.length) {
            case 1:
                return this.constructor$1(...args$);
            case 2:
                if (args$[0].constructor.name === 'LineOp') {
                    return this.constructor$2_1(...args$);
                } else {
                    return this.constructor$2_2(...args$);
                }
            case 3:
                return this.constructor$3(...args$);
            case 5:
                return this.constructor$5(...args$);
        }
    }

    constructor$1(singleLine) {
        this.lineOne = singleLine;
        this.type = LineGroup.Single;
        this.lines.push(this.lineOne);
    }

    constructor$2_1(leftOp, line) {
        this.lineOp = leftOp;
        this.lineOne = line;
        this.type = LineGroup.Prefix;
        this.lines.push(this.lineOp);
        this.lines.push(this.lineOne);

        if (leftOp.getKey() === Key.Inc || leftOp.getKey() === Key.Dec) {
            if (line.getLastCall() == null || (line.getLastCall().getType() !== LineCall.Field &&
                line.getLastCall().getType() !== LineCall.IndexCall)) {
                this.getParent().error(leftOp.getToken(), Error.lineRefOperator);
            }
        }
    }

    constructor$2_2(line, rightOp) {
        this.lineOp = rightOp;
        this.lineOne = line;
        this.type = LineGroup.Postfix;
        this.lines.push(this.lineOne);
        this.lines.push(this.lineOp);
        if (rightOp.getKey() === Key.Inc || rightOp.getKey() === Key.Dec) {
            if (line.getLastCall() === null || (line.getLastCall().getType() !== LineCall.Field && line.getLastCall().getType() !== LineCall.IndexCall)) {
                this.getParent().error(rightOp.getToken(), Error.lineRefOperator);
            }
        }
    }

    constructor$3(leftLine, lineOp, rightLine) {
        this.lineOne = leftLine;
        this.lineOp = lineOp;
        this.lineTwo = rightLine;
        this.type = LineGroup.Middle;
        this.lines.push(this.lineOne);
        this.lines.push(this.lineOp);
        this.lines.push(this.lineTwo);
        if (this.lineOp.getKey().op === LineOp.SetPrecedence) {
            if (leftLine.getLastCall() === null || (leftLine.getLastCall().getType() !== LineCall.Field && leftLine.getLastCall().getType() !== LineCall.IndexCall)) {
                this.getParent().error(this.lineOp.getToken(), Error.lineSetOperator);
            }
        }
    }

    constructor$5(leftLine, lineOp, rightLine, lineElse, lineTree) {
        this.lineOne = leftLine;
        this.lineOp = lineOp;
        this.lineTwo = rightLine;
        this.lineElse = lineElse;
        this.lineTree = lineTree;
        this.type = LineGroup.Ternary;
        this.lines.push(leftLine);
        this.lines.push(this.lineOp);
        this.lines.push(this.lineTwo);
        this.lines.push(this.lineElse);
        this.lines.push(this.lineTree);
    }

    getParent() {
        return this.parent;
    }

    getType() {
        return this.type;
    }

    getLineOne() {
        return this.lineOne;
    }

    getLineTwo() {
        return this.lineTwo;
    }

    getLineTree() {
        return this.lineTree;
    }

    getLineOp() {
        return this.lineOp;
    }

    getLineElse() {
        return this.lineElse;
    }

    getLines() {
        return this.lines;
    }

    getReturnType() {
        if (this.type === LineGroup.Middle && this.lineOp !== null) {
            if (this.lineOp.getKey() === Key.And || this.lineOp.getKey() === Key.Or) {
                return LineValue.Boolean;
            } else if (this.lineOp.getPrecedence() === Key.Set.op) {
                return LineValue.Void;
            }
        }
        return LineValue.Object;
    }

    toString() {
        if (this.type === LineGroup.Single) {
            return '(' + this.lineOne + ')';
        }
        if (this.type === LineGroup.Prefix) {
            return '(' + this.lineOp + ' ' + this.lineOne + ')';
        }
        if (this.type === LineGroup.Postfix) {
            return '(' + this.lineOne + ' ' + this.lineOp + ')';
        }
        if (this.type === LineGroup.Middle) {
            return '(' + this.lineOne + ' ' + this.lineOp + ' ' + this.lineTwo + ')';
        }
        if (this.type === LineGroup.Ternary) {
            return '(' + this.lineOne + ' ' + this.lineOp + ' ' + this.lineTwo + ' ' + this.lineElse + ' ' + this.lineTree + ')';
        }
        return 'empty line';
    }
}
LineGroup.Single = 'Single';
LineGroup.Prefix = 'Prefix';
LineGroup.Postfix = 'Postfix';
LineGroup.Middle = 'Middle';
LineGroup.Ternary = 'Ternary';

module.exports = LineGroup;
Key = require("simplex/lang/lexer/Key.js");
LineOp = require("simplex/lang/parser/logic/line/LineOp.js");
LineCall = require("simplex/lang/parser/logic/line/call/LineCall.js");
Error = require("simplex/lang/parser/logic/error/Error.js");